src/Markdown/Block.elm

```diff
@@ -41,48 +58,177 @@ In the simplest case, you can pass this directly to a renderer:
 
     main : Html msg
     main =
-        case astResult of
-            Ok ast ->
-                Markdown.Parser.renderDefaultHtml ast
+        case
+            astResult
+                |> Result.mapError deadEndsToString
+                |> Result.andThen (\ast -> Markdown.Renderer.render Markdown.Renderer.defaultHtmlRenderer ast)
+        of
+            Ok rendered ->
+                div [] rendered
 
             Err errors ->
-                Html.text "Encountered parsing errors."
+                text errors
 
 -}
 type Block
-    = Heading Int (List Inline)
-    | Body (List Inline)
-    | Html String (List Attribute) (List Block)
-    | UnorderedListBlock
-        (List
-            { task : Maybe Bool
-            , body : List Inline
-            }
-        )
-    | OrderedListBlock Int (List (List Inline))
-    | CodeBlock Markdown.CodeBlock.CodeBlock
-    | ThematicBreak
+    = -- Container Blocks
+      HtmlBlock (Html Block)
+    | UnorderedList (List (ListItem Inline))
+    | OrderedList Int (List (List Inline))
     | BlockQuote (List Block)
+      -- Leaf Blocks With Inlines
+    | Heading HeadingLevel (List Inline)
+    | Paragraph (List Inline)
+      -- Table TODO  https://github.github.com/gfm/#tables-extension-
+      -- Leaf Blocks Without Inlines
+    | CodeBlock { body : String, language : Maybe String }
+    | ThematicBreak
+
+
+{-| The value for an unordered list item, which may contain a task.
+-}
+type ListItem children
+    = ListItem Task (List children)
+
+
+{-| A task (or no task), which may be contained in a ListItem.
+-}
+type Task
+    = NoTask
+    | IncompleteTask
+    | CompletedTask
+
+
+{-| Markdown limits headings to level 6 or less. HTML does this, too (`<h7>` is not supported by most browsers).
+This type represents the possible heading levels that a Markdown heading block may have.
+
+If you do a heading level above 6, the `#` characters will be treated as literal `#`'s.
+
+-}
+type HeadingLevel
+    = H1
+    | H2
+    | H3
+    | H4
+    | H5
+    | H6
```

Comment:
Now that HeadingLevel no longer is an integer, I would appreciate if there was a function bumpHeadingLevel : HeadingLevel -> HeadingLevel, as I was simply using + 1 before, and I assume other might have been doing that before.

When testing this branch I already had a definition like this:
```elm
bumpHeadingLevel : HeadingLevel -> HeadingLevel
bumpHeadingLevel level =
    case level of
        H1 ->
            H2

        H2 ->
            H3

        H3 ->
            H4

        H4 ->
            H5

        H5 ->
            H6

        H6 ->
            H6
```

Reply:
Hey! Thanks for the feedback @matheus23!

I think this is a great idea for a function, but I don't want to clutter the core modules too much with helpers like this. Maybe once the API stabilizes, we could talk about a separate package for helpers like this?

There are so many transformations that could be defined, I think that adding these types of things to the core library could be a never-ending thing. So I think that this is probably a great place for an ecosystem of transformer plugins, much like [RemarkJS has tons of plugins defined separately from the core package](https://github.com/remarkjs/remark/blob/master/doc/plugins.md#list-of-plugins).

What do you think of that idea?

Reply:
Absolutely! I'm in the process of writing my own already! :)